resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -407,38 +415,42 @@ private[spark] class KubernetesClusterSchedulerBackend(scheduler: TaskSchedulerI
           case Some(pod) =>
             localFailedPods.get(pod.getMetadata.getName) match {
               case Some(executorExited: ExecutorExited) =>
-                removeExecutor(executorId, executorExited)
                 logDebug(s"Removing executor $executorId with loss reason "
                   + executorExited.message)
+                removeExecutor(executorId, executorExited)
                 if (!executorExited.exitCausedByApp) {
-                  removedExecutors.add(executorId)
+                  executorsToRecover.add(executorId)
                 }
               case None =>
-                val checkedAttempts = executorAttempts.getOrElse(executorId, 0)
-                executorAttempts.put(executorId, checkedAttempts + 1)
+                removeExecutorOrIncrementLossReasonCheckCount(executorId)
             }
           case None =>
-            val checkedAttempts = executorAttempts.getOrElse(executorId, 0)
-            executorAttempts.put(executorId, checkedAttempts + 1)
+            removeExecutorOrIncrementLossReasonCheckCount(executorId)
         }
       }
 
-      for ((executorId, attempts) <- executorAttempts) {
-        if (attempts >= MAX_ATTEMPTS) {
-          removeExecutor(executorId, SlaveLost("Executor lost for unknown reasons"))
-          removedExecutors.add(executorId)
-        }
-      }
-      removedExecutors.foreach(executorId =>
+      executorsToRecover.foreach(executorId =>
         EXECUTORS_TO_REMOVE_LOCK.synchronized {
           executorsToRemove -= executorId
           executorAttempts -= executorId
         }
       )
-      if (removedExecutors.nonEmpty) {
-        requestExecutors(removedExecutors.size)
+      if (executorsToRecover.nonEmpty &&
+        recoveredExecutorCount < MAX_ALLOWED_EXECUTOR_RECOVERY_ATTEMPTS) {
+        requestExecutors(executorsToRecover.size)
```

Comment:
Compute numExecutorsToRecover as Math.min(executorsToRecover.size, MAX_ALLOWED_EXECUTOR_RECOVERY_ATTEMPTS - recoveredExecutorCount) and pass it to requestExecutors so that we don't go above the max?